src/stest/src/file.rs

```diff
@@ -0,0 +1,232 @@
+        }
+        result(&self, file)
+            .unwrap()
+    }
+
+    pub fn is_pipe(&self) -> bool {
+        fn is_pipe(metadata: Metadata) -> bool {
+            metadata.file_type()
+                .is_fifo()
+        }
+        let result = self.path_buf
+            .metadata();
+        result.map(is_pipe)
+            .unwrap()
+    }
+
+    /// Check if a unix file has any readable bit set (user, group, or other).
+    /// See: https://en.wikipedia.org/wiki/File-system_permissions#Numeric_notation
+    /// See: https://en.wikipedia.org/wiki/Bitwise_operation#AND
+    pub fn is_readable(&self) -> bool {
+        fn is_readable(mode: u32) -> bool { mode & 0o444 != 0 }
+        let result = self.mode();
+        result.map(is_readable)
+            .unwrap()
+    }
+
+    pub fn has_size_greater_than_zero(&self) -> bool {
+        fn has_size_greater_than_zero(metadata: Metadata) -> bool {
+            let len = metadata.len();
+            len > 0
+        }
+        let result = self.metadata();
+        result.map(has_size_greater_than_zero)
+            .unwrap()
+    }
+
+    /// Check if the file has the set-user-ID bit set.
+    /// See: https://stackoverflow.com/a/50045872/8732788
+    /// See: https://en.wikipedia.org/wiki/Setuid
+    pub fn has_set_user_id(&self) -> bool {
+        fn has_set_user_id(mode: u32) -> bool { mode & 0o4000 != 0 }
+        let result = self.mode();
+        result.map(has_set_user_id)
+            .unwrap()
+    }
+
+    /// Check if the file has any writable bit set (user, group, or other).
+    /// See: https://en.wikipedia.org/wiki/File-system_permissions#Numeric_notation
+    /// See: https://en.wikipedia.org/wiki/Bitwise_operation#AND
+    pub fn is_writable(&self) -> bool {
+        fn is_writable(mode: u32) -> bool { mode & 0o222 != 0 }
+        let result = self.mode();
+        result.map(is_writable)
+            .unwrap()
+    }
+
+    /// Check if the file has any executable bit set (user, group, or other).
+    /// See: https://en.wikipedia.org/wiki/File-system_permissions#Numeric_notation
+    /// See: https://en.wikipedia.org/wiki/Bitwise_operation#AND
+    pub fn is_executable(&self) -> bool {
+        fn is_executable(mode: u32) -> bool { mode & 0o111 != 0 }
+        let result = self.mode();
+        result.map(is_executable)
+            .unwrap()
+    }
+
+    fn metadata(&self) -> Result<Metadata, IOError> {
+        self.path_buf
+            .metadata()
+    }
+
+    fn mode(&self) -> Result<u32, IOError> {
+        fn metadata_to_mode(metadata: Metadata) -> u32 {
+            metadata.permissions()
+                .mode()
+        }
+        self.metadata()
+            .map(metadata_to_mode)
+    }
+
+}
+
+impl Display for File {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        let display = self.path_buf.display();
+        write!(f, "{display}")
+    }
+}
+
+impl FromStr for File {
+    type Err = std::convert::Infallible;
```

Comment:
Why is this `std::convert::Infallible`?

Reply:
This seemed strange to me too, but apparently that is [the `Err` type for `FromStr` on `PathBuf`][1]. I guess you can `PathBuf.from_str(str)` always succeeds?

[1]: https://doc.rust-lang.org/src/std/path.rs.html#1705-1712